core/src/main/java/org/apache/accumulo/core/client/ClientThreadPools.java

```diff
@@ -0,0 +1,136 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.client;
+
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadPoolExecutor;
+
+import org.apache.accumulo.core.clientImpl.ClientContext;
+
+public interface ClientThreadPools {
+
+  /**
+   * return a shared scheduled executor for trivial tasks
+   *
+   * @param ctx
+   * @return ScheduledThreadPoolExecutor
+   */
+  ScheduledThreadPoolExecutor getSharedScheduledExecutor(ClientContext context);
+
+  /**
+   * ThreadPoolExecutor that runs bulk import tasks
+   *
+   * @param ctx
+   * @param numThreads
+   * @return ThreadPoolExecutor
+   */
+  ThreadPoolExecutor getBulkImportThreadPool(ClientContext ctx, int numThreads);
+
+  /**
+   * ThreadPoolExecutor that runs tasks to contact Compactors to get running compaction information
+   *
+   * @param numThreads
+   * @return ThreadPoolExecutor
+   */
+  ThreadPoolExecutor getExternalCompactionActiveCompactionsPool(ClientContext ctx, int numThreads);
+
+  /**
+   * ThreadPoolExecutor used for fetching data from the TabletServers
+   *
+   * @param ctx
+   * @return ThreadPoolExecutor
+   */
+  ThreadPoolExecutor getScannerReadAheadPool(ClientContext ctx);
+
+  /**
+   * ThreadPoolExecutor used for adding splits to a table
+   *
+   * @param ctx
+   * @return ThreadPoolExecutor
+   */
+  ThreadPoolExecutor getAddSplitsThreadPool(ClientContext ctx);
+
+  /**
+   * ThreadPoolExecutor used for fetching data from the TabletServers
+   *
+   * @param ctx
+   * @param numQueryThreads
+   * @param batchReaderInstance
+   * @return ThreadPoolExecutor
+   */
+  ThreadPoolExecutor getBatchReaderThreadPool(ClientContext ctx, int numQueryThreads,
+      int batchReaderInstance);
+
+  /**
+   * ScheduledThreadPoolExecutor that runs tasks for the BatchWriter to meet the users latency
+   * goals.
+   * 
+   * @param ctx
+   *          client context object
+   * @return ScheduledThreadPoolExecutor
+   */
+  ScheduledThreadPoolExecutor getBatchWriterLatencyTasksThreadPool(ClientContext ctx);
+
+  /**
+   * ThreadPoolExecutor that runs the tasks of binning mutations
+   * 
+   * @param ctx
+   *          client context object
+   * @return ThreadPoolExecutor
+   */
+  ThreadPoolExecutor getBatchWriterBinningThreadPool(ClientContext ctx);
+
+  /**
+   * ThreadPoolExecutor that runs the tasks of sending mutations to TabletServers
+   *
+   * @param ctx
+   *          client context object
+   * @return ThreadPoolExecutor
+   */
+  ThreadPoolExecutor getBatchWriterSendThreadPool(ClientContext ctx, int numSendThreads);
+
+  /**
+   * ThreadPoolExecutor that runs clean up tasks when close is called on the ConditionalWriter
+   *
+   * @param ctx
+   * @return ThreadPoolExecutor
+   */
+  ThreadPoolExecutor getConditionalWriterCleanupTaskThreadPool(ClientContext ctx);
```

Comment:
Having one central location that entagles itself with so many disparate public API methods seems less than ideal. I think it would be better if each entry point (batch scanner, batch writer, conditional writer, etc.) added an option to their existing builders/config to set a thread pool... or more narrowly, just an uncaught exception handler for the thread pools we manage internally.

Reply:
I was trying to reduce the number of places users would have to modify their pre-2.1.0 code. This way would allow them to create their own implementation when they create the client instead of specifying for each scanner/writer/etc. Additionally, by creating their own ThreadPool's, they can set their own uncaught exception handler.

Reply:
We did try to make those builders/config extensible, so they wouldn't have to modify anything unless they were taking advantage of the new feature... However, I think I have a better idea anyway.

I noticed that this class is essentially a ThreadPoolExecutorFactory for either `ThreadPoolExecutor` or `ScheduledThreadPoolExecutor`. Instead of having so many disparate methods, we can make the interface substantially simpler, and still support all the different uses, by passing in some kind of context/scope, as in:

```java
    // this name is bad, but illustrates the idea; this could even be a String, if we don't want to constrain it
    enum TheadPoolScope {
      BATCH_WRITER, BATCH_SCANNER, CONDITIONAL_WRITER;
    }
    ThreadPoolExecutor getExecutor(ThreadPoolScope scope, ConfigSource conf);
    ScheduledThreadPoolExecutor getScheduledThreadPoolExecutor(ThreadPoolScope scope, ConfigSource conf);
```

Reply:
see 2effcce.

Reply:
> The main idea here is to keep the API as free of bloat as possible

It would be nice to trim this down. Not sure how, but all of the enums do seem like a bit much.

Reply:
I'm not sure how we do that. We need to pass enough information to determine which pool to return and how to configure it. I'm open to suggestions.

Reply:
The enums do not properly capture all of the complexity.  For example I see for one enum a SynchronousQueue is used and for another enum 0 to 3 threads are used.  It seems like the enum could be dropped and everything that is needed for creation pushed into the config object, maybe that becomes to complex for anyone to ever use properly though.

Reply:
I have some changes locally where I'm cleaning up potentially unclosed thread pools (for example, if someone did not call close). This means that clients might be able to specify their own implementation, but they would not be able to share the thread pool within the client or with the clients application. I'm starting to think that the only thing a user would be able to configure is the number of threads in the pools. Currently the number of threads is either hardcoded or based on a property.